### ISSUE TITLE
Expose context for acorn `SyntaxError`s

### ISSUE THREAD

**Issue by cspotcode** [pugjs/pug#2830]
**Pug Version:** 2.0.0-rc.2

**Node Vesrion:** v8.1.2

When an acorn parsing error occurs, a SyntaxError is thrown from acorn.  I can catch this error in my code, but I can't associate it with the JavaScript source that caused the error.  For example:

```js
pug.compile(`
// lots of pug here...
-
    Object.assign(biff, {
        foo // I forgot the comma.  Whoops!
        bar
    });
`);
```

Acorn [exposes `pos` and `loc` values on the error](https://github.com/ternjs/acorn#main-parser) that tell me where in the code the SyntaxError occurred, but pug does not expose the string of JavaScript that's being parsed.

I propose that pug catch thrown acorn SyntaxErrors and either wrap or augment them to reveal the code that caused the syntax error.  This can be as simple as `if(isAcornSyntaxError(e)) e.code = stringOfJsCode;` because then my code can inspect `e.code`.

However, a more helpful solution is to modify the error's message to include a few lines of context from the JS source being parsed.  I think this can be accomplished using the same code that Pug uses to show context within Pug source code.

**Comment by TimothyGu** [comment]
We already do that with several other syntactical elements in Pug, like interpolation:

```pug
| #{Object.assign(biff, { foo bar })}
```

```
index.pug:1:31
  > 1| | #{Object.assign(biff, { foo bar })}
-------------------------------------^

Syntax Error: Unexpected token
```

The challenge for `-` is that the expression may not be valid standalone JS. In other words, this piece of Pug code is completely valid:

```pug
- for (let i = 0; i++; i < 10) {
| abc
- }
```

There is no way for us to detect which cases of invalid JS are intended, and which are not.

The SyntaxError you are seeing happens when we are parsing the *generated* JS. So far we haven't implemented source mapping from generated JS to Pug yet.

**Comment by cspotcode** [comment]
Yeah I know, but showing context in those situations is still helpful; way
more than the current state of affairs.  Otherwise I get a syntax error
with zero clue where it occurred.  Giving me 7 or so lines of context at
least lets me grep my codebase for the mistake rather than being completely
blind.  The fact is that when I'm getting an error, anything that makes it
less cryptic is an improvement.

But you raise a good point; the error message should indicate that the
syntax error occurred in generated code and maybe hide the line numbers
since they won't be meaningful.

I know that proper source mapping is the ideal implementation, but until
that lands, I can send you a 5-line PR that offers a big productivity boost
today.  I wrote it right after raising this issue.

**Comment by TimothyGu** [comment]
> I can send you a 5-line PR that offers a big productivity boost today.

That can be helpful, please do.

**Comment by cspotcode** [comment]
Ok, will do.  I noticed that [pugjs/with](https://github.com/pugjs/with) has been rewritten to use babel instead of acorn but a new version hasn't been published.  Some of the parsing errors that must be caught are thrown when `with` parses code, meaning I'll need to send 2 PRs: one for `with` and one for `pug`.  Should I write them against the new babel-based `with` implementation?

**Comment by ForbesLindesay** [comment]
Yes, we can update to the new version of with once you've added what you need to for better error messages.

**Comment by Dan503** [comment]
I'm wondering if Pug is able to detect when a block of JS has started and ended.

If you detected the start of a JS block, but then never detected the end the block, it would be due to acorn throwing an error. You could then possibly indicate which block of js and in which file the error occurred in that way.

I'm thinking it would sort of work in the same sort of way to how Gulp 4 works when it throws errors from tasks not signalling when they are done.

Or is that not really possible with how Pug implements JS blocks?

Or maybe do something with Try/Catch?

`````pug
-
    //Detects a code block has started, initiates a "try" statement
    Object.assign({}, {
        foo // I forgot the comma.  Whoops!
        bar
    });
    //Caught an error, log the file and line number that block started at
`````

It's really hard to debug JS block errors right now :(

**Comment by ForbesLindesay** [comment]
We do not currently track where blocks start and end.  This is because we allow for things like:

```pug
- try {
div= thisMethodThrowsAnError()
- } catch (ex) {
div something went wrong
- }
```

I would be happy to add something that tracks this better and would be able to catch examples like yours (i.e. the brackets match up, but there is some other syntax error).  Unfortunately, I don't have a lot of time to work on pug, and my current focus is stabilising the 2.0.0 release and improving documentation.  If you want me to have more time to work on pug, please consider contributing to pug's funding https://opencollective.com/pug (or persuade other people you know to contribute).

**Comment by Dan503** [comment]
I didn't know that even existed. O_o

You may want to place that link in a more prominent place on the website.

If you put a page on pugjs.org called "Support Pug" or something with a little blurb and a link to the site then your revenue would likely grow substantially.

People love your creation but they have no idea how to support it.

**Comment by ForbesLindesay** [comment]
I'm working on it :) expect it to become gradually more prominent over the next couple of months.
